File: solutions/python/heap-priority-queue/last-stone-weight.py
# LeetCode 1046 — Last Stone Weight (Easy)
# Category: Heap / Priority Queue · Approach: Max-heap (negated)
# Time: O(n log n) | Space: O(n)
# Source: https://leetcode.com/problems/last-stone-weight/

import heapq

def last_stone_weight(stones: list[int]) -> int:
    heap = [-s for s in stones]
    heapq.heapify(heap)
    while len(heap) > 1:
        y = -heapq.heappop(heap)
        x = -heapq.heappop(heap)
        if y != x:
            heapq.heappush(heap, -(y - x))
    return -heap[0] if heap else 0

File: solutions/typescript/heap-priority-queue/last-stone-weight.ts
// LeetCode 1046 — Last Stone Weight (Easy)
// Category: Heap / Priority Queue · Approach: Sort each round (clear)
// Time: O(n log n) | Space: O(n)
// Source: https://leetcode.com/problems/last-stone-weight/

function lastStoneWeight(stones: number[]): number {
  const arr = [...stones];
  while (arr.length > 1) {
    arr.sort((a, b) => a - b);
    const y = arr.pop()!;
    const x = arr.pop()!;
    if (y !== x) arr.push(y - x);
  }
  return arr.length ? arr[0] : 0;
}
